src/Elmah.Io.AspNetCore/MessageShipper.cs

```diff
@@ -11,59 +12,66 @@ namespace Elmah.Io.AspNetCore
+        public static void Ship(Exception exception, string title, HttpContext context, ElmahIoOptions options, IBackgroundTaskQueue queue)
         {
-            var baseException = exception?.GetBaseException();
-            var createMessage = new CreateMessage
+            queue.QueueBackgroundWorkItem(async token =>
             {
-                DateTime = DateTime.UtcNow,
-                Detail = Detail(exception, options),
-                Type = baseException?.GetType().Name,
-                Title = title,
-                Data = exception?.ToDataList(),
-                Cookies = Cookies(context),
-                Form = Form(context),
-                Hostname = context.Request?.Host.Host,
-                ServerVariables = ServerVariables(context),
-                StatusCode = StatusCode(exception, context),
-                Url = context.Request?.Path.Value,
-                QueryString = QueryString(context),
-                Method = context.Request?.Method,
-                Severity = Severity(exception, context),
-                Source = Source(baseException),
-            };
-
-            TrySetUser(context, createMessage);
-
-            if (options.OnFilter != null && options.OnFilter(createMessage))
-            {
-                return;
-            }
-
-            var elmahioApi = new ElmahioAPI(new ApiKeyCredentials(options.ApiKey), new HttpClientHandler
-            {
-                UseProxy = options.WebProxy != null,
-                Proxy = options.WebProxy,
+                var baseException = exception?.GetBaseException();
+                var createMessage = new CreateMessage
+                {
+                    DateTime = DateTime.UtcNow,
+                    Detail = Detail(exception, options),
+                    Type = baseException?.GetType().Name,
+                    Title = title,
+                    Data = exception?.ToDataList(),
+                    Cookies = Cookies(context),
+                    Form = Form(context),
+                    Hostname = context.Request?.Host.Host,
+                    ServerVariables = ServerVariables(context),
+                    StatusCode = StatusCode(exception, context),
+                    Url = context.Request?.Path.Value,
+                    QueryString = QueryString(context),
+                    Method = context.Request?.Method,
+                    Severity = Severity(exception, context),
+                    Source = Source(baseException),
+                    Application = options.Application,
+                };
+
+                TrySetUser(context, createMessage);
+
+                if (options.OnFilter != null && options.OnFilter(createMessage))
+                {
+                    return;
+                }
+
+                var elmahioApi = new ElmahioAPI(new ApiKeyCredentials(options.ApiKey), new HttpClientHandler
```

Comment:
it probably just creates a new one.. :D
I think dns is an issue, since this client library has no control over when the clients are restarted (which would clear the dns cache). In theory, it could be years :D
I would go with the static one, and then create a new one if the old one is 1 hour old or something. Maybe also the proxy change thing. That allows you to have more flexibility in changing your elmah backend dns.

Reply:
I think that I'll need to register `ElmahioAPI` as a singleton when used from ASP.NET Core. As I can see in the documentation it is recommended to share classes extending `ServiceClient` (since that contains a `HttpClient` internally).